base/gfspvgmgr/virtual_group_manager.go

```diff
@@ -482,3 +504,171 @@ func (vgm *virtualGroupManager) releaseSPAndGVGLoop() {
 		}
 	}
 }
+
+type HealthChecker struct {
+	mutex sync.RWMutex
+
+	sps          map[uint32]*sptypes.StorageProvider // all sps,  id -> StorageProvider
+	unhealthySPs map[uint32]*sptypes.StorageProvider
+
+	chainClient consensus.Consensus // query VG params from chain
+}
+
+func NewHealthChecker(chainClient consensus.Consensus) *HealthChecker {
+	sps := make(map[uint32]*sptypes.StorageProvider)
+	unhealthySPs := make(map[uint32]*sptypes.StorageProvider)
+	return &HealthChecker{sps: sps, unhealthySPs: unhealthySPs, chainClient: chainClient}
+}
+
+func (checker *HealthChecker) addSP(sp *sptypes.StorageProvider) {
+	checker.mutex.Lock()
+	defer checker.mutex.Unlock()
+
+	if !sp.IsInService() {
+		return
+	}
+	if _, exists := checker.sps[sp.GetId()]; !exists {
+		checker.sps[sp.GetId()] = sp
+	}
+}
+
+func (checker *HealthChecker) isSPHealthy(spID uint32) bool {
+	checker.mutex.RLock()
+	defer checker.mutex.RUnlock()
+
+	if sp, exists := checker.sps[spID]; exists {
+		if _, unhealthyExists := checker.unhealthySPs[spID]; unhealthyExists {
+			log.CtxDebugw(context.Background(), "the sp is treated as unhealthy", "sp", sp)
+			return false
+		} else {
+			return true
+		}
+	}
+	log.CtxDebugw(context.Background(), "the sp isn't exist in sps map, is treated as unhealthy", "sps", checker.sps)
```

Comment:
In what scenario would this case occur? Is it safer to treat it as a healthy SP?